posthog/session_recordings/synthetic_playlists.py

```diff
@@ -254,6 +288,311 @@ def to_synthetic_playlist(self) -> "SyntheticPlaylistDefinition":
+                AND min_first_timestamp >= %(history_start)s
+                AND url != ''
+            GROUP BY session_id, url
+            ORDER BY first_seen DESC
+            LIMIT 50000
+        """
+
+        result = sync_execute(
+            query,
+            {
+                "team_id": team.pk,
+                "history_start": history_window_start,
+            },
+        )
+
+        # Build pattern tracking data structures
+        pattern_first_seen: dict[str, datetime] = {}
+        pattern_sessions: dict[str, set[str]] = {}  # pattern -> set of session_ids
+
+        for session_id, raw_url, first_seen_ts in result:
+            normalized = NewUrlsSyntheticPlaylistSource._normalize_url(raw_url)
+
+            # Track earliest appearance of this pattern
+            if normalized not in pattern_first_seen or first_seen_ts < pattern_first_seen[normalized]:
+                pattern_first_seen[normalized] = first_seen_ts
+
+            # Track sessions for patterns in the lookback window (for counting)
+            if lookback_start <= first_seen_ts <= now:
+                if normalized not in pattern_sessions:
+                    pattern_sessions[normalized] = set()
+                pattern_sessions[normalized].add(session_id)
+
+        # Filter to patterns that FIRST appeared within the lookback window
+        new_patterns_with_counts = {
+            pattern: len(pattern_sessions.get(pattern, set()))
+            for pattern, first_seen in pattern_first_seen.items()
+            if lookback_start <= first_seen <= now
+        }
+
+        # Sort by count (descending) then by pattern name, take top 20
+        sorted_patterns = sorted(
+            new_patterns_with_counts.items(),
+            key=lambda x: (-x[1], x[0]),  # Sort by count desc, then pattern asc
+        )[:20]
+
+        result_dict = dict(sorted_patterns)
+
+        cache.set(cache_key, result_dict, NewUrlsSyntheticPlaylistSource.CACHE_TTL)
+        return result_dict
+
+    @staticmethod
+    def _get_new_urls(team: Team) -> list[str]:
+        """
+        Get just the list of new URL patterns (for backwards compatibility).
+        """
+        return list(NewUrlsSyntheticPlaylistSource._get_new_urls_with_counts(team).keys())
+
+    def get_session_ids(self, team: Team, user: User, limit: int | None = None, offset: int | None = None) -> list[str]:
+        """
+        Get session IDs for sessions that visited URLs matching this pattern.
+        Since self.url is a normalized pattern (like /project/{id}/settings),
+        we need to fetch all URLs and match them after normalization.
+
+        Uses caching to avoid expensive queries on every request.
+        """
+        if not self.url:
+            return []
+
+        # Cache key includes the pattern URL to ensure different patterns have different caches
+        cache_key = f"{self.CACHE_KEY_PREFIX}_sessions_{team.pk}_{hashlib.sha256(self.url.encode()).hexdigest()[:16]}"
```

Comment:
i wonder if we could populate this cache when we're generating the session to pattern lists
to avoid the lookup here 

not blocking... just wondering how this all will perform against prod :))

---

and repeating myself cos that might annoy me into fixing it, we do need to separate things so we can pass lists of session ids to the normal listing code the API uses so we get the TTL expiry checks and pagination etc that that offers

but that's not a problem of this PR it's this synthetic collection and timed collection counting stuff generally should be doing it